README
NAME
    IPC::Run - system() and background procs w/ piping, redirs, ptys (Unix,
    Win32)

SYNOPSIS
       ## First,a command to run:
          my @cat = qw( cat );

       ## Using run() instead of system():
          use IPC::Run qw( run timeout );

          run \@cat, \$in, \$out, \$err, timeout( 10 ) or die "cat: $?"

          # Can do I/O to sub refs and filenames, too:
          run \@cat, '<', "in.txt", \&out, \&err or die "cat: $?"
          run \@cat, '<', "in.txt", '>>', "out.txt", '2>>', "err.txt";


          # Redirecting using pseudo-terminals instead of pipes.
          run \@cat, '<pty<', \$in,  '>pty>', \$out_and_err;

       ## Scripting subprocesses (like Expect):

          use IPC::Run qw( start pump finish timeout );

          # Incrementally read from / write to scalars. 
          # $in is drained as it is fed to cat's stdin,
          # $out accumulates cat's stdout
          # $err accumulates cat's stderr
          # $h is for "harness".
          my $h = start \@cat, \$in, \$out, \$err, timeout( 10 );

          $in .= "some input\n";
          pump $h until $out =~ /input\n/g;

          $in .= "some more input\n";
          pump $h until $out =~ /\G.*more input\n/;

          $in .= "some final input\n";
          finish $h or die "cat returned $?";

          warn $err if $err; 
          print $out;         ## All of cat's output

       # Piping between children
          run \@cat, '|', \@gzip;

       # Multiple children simultaneously (run() blocks until all
       # children exit, use start() for background execution):
          run \@foo1, '&', \@foo2;

       # Calling \&set_up_child in the child before it executes the
       # command (only works on systems with true fork() & exec())
       # exceptions thrown in set_up_child() will be propagated back
       # to the parent and thrown from run().
          run \@cat, \$in, \$out,
             init => \&set_up_child;

       # Read from / write to file handles you open and close
          open IN,  '<in.txt'  or die $!;
          open OUT, '>out.txt' or die $!;
          print OUT "preamble\n";
          run \@cat, \*IN, \*OUT or die "cat returned $?";
          print OUT "postamble\n";
          close IN;
          close OUT;

       # Create pipes for you to read / write (like IPC::Open2 & 3).
          $h = start
             \@cat,
                '<pipe', \*IN,
                '>pipe', \*OUT,
                '2>pipe', \*ERR 
             or die "cat returned $?";
          print IN "some input\n";
          close IN;
          print <OUT>, <ERR>;
          finish $h;

       # Mixing input and output modes
          run \@cat, 'in.txt', \&catch_some_out, \*ERR_LOG );

       # Other redirection constructs
          run \@cat, '>&', \$out_and_err;
          run \@cat, '2>&1';
          run \@cat, '0<&3';
          run \@cat, '<&-';
          run \@cat, '3<', \$in3;
          run \@cat, '4>', \$out4;
          # etc.

       # Passing options:
          run \@cat, 'in.txt', debug => 1;

       # Call this system's shell, returns TRUE on 0 exit code
       # THIS IS THE OPPOSITE SENSE OF system()'s RETURN VALUE
          run "cat a b c" or die "cat returned $?";

       # Launch a sub process directly, no shell.  Can't do redirection
       # with this form, it's here to behave like system() with an
       # inverted result.
          $r = run "cat a b c";

       # Read from a file in to a scalar
          run io( "filename", 'r', \$recv );
          run io( \*HANDLE,   'r', \$recv );

DESCRIPTION
    IPC::Run allows you to run and interact with child processes using
    files, pipes, and pseudo-ttys. Both system()-style and scripted usages
    are supported and may be mixed. Likewise, functional and OO API styles
    are both supported and may be mixed.

    Various redirection operators reminiscent of those seen on common Unix
    and DOS command lines are provided.

    Before digging in to the details a few LIMITATIONS are important enough
    to be mentioned right up front:

    Win32 Support
        Win32 support is working but EXPERIMENTAL, but does pass all
        relevant tests on NT 4.0. See "Win32 LIMITATIONS".

    pty Support
        If you need pty support, IPC::Run should work well enough most of
        the time, but IO::Pty is being improved, and IPC::Run will be
        improved to use IO::Pty's new features when it is release.

        The basic problem is that the pty needs to initialize itself before
        the parent writes to the master pty, or the data written gets lost.
        So IPC::Run does a sleep(1) in the parent after forking to
        (hopefully) give the child a chance to run. This is a kludge that
        works well on non heavily loaded systems :(.

        ptys are not supported yet under Win32, but will be emulated...

    Debugging Tip
        You may use the environment variable "IPCRUNDEBUG" to see what's
        going on under the hood:

           $ IPCRUNDEBUG=basic   myscript     # prints minimal debugging
           $ IPCRUNDEBUG=data    myscript     # prints all data reads/writes
           $ IPCRUNDEBUG=details myscript     # prints lots of low-level details
           $ IPCRUNDEBUG=gory    myscript     # (Win32 only) prints data moving through
                                              # the helper processes.

    We now return you to your regularly scheduled documentation.

  Harnesses
    Child processes and I/O handles are gathered in to a harness, then
    started and run until the processing is finished or aborted.

  run() vs. start(); pump(); finish();
    There are two modes you can run harnesses in: run() functions as an
    enhanced system(), and start()/pump()/finish() allow for background
    processes and scripted interactions with them.

    When using run(), all data to be sent to the harness is set up in
    advance (though one can feed subprocesses input from subroutine refs to
    get around this limitation). The harness is run and all output is
    collected from it, then any child processes are waited for:

       run \@cmd, \<<IN, \$out;
       blah
       IN

       ## To precompile harnesses and run them later:
       my $h = harness \@cmd, \<<IN, \$out;
       blah
       IN

       run $h;

    The background and scripting API is provided by start(), pump(), and
    finish(): start() creates a harness if need be (by calling harness())
    and launches any subprocesses, pump() allows you to poll them for
    activity, and finish() then monitors the harnessed activities until they
    complete.

       ## Build the harness, open all pipes, and launch the subprocesses
       my $h = start \@cat, \$in, \$out;
       $in = "first input\n";

       ## Now do I/O.  start() does no I/O.
       pump $h while length $in;  ## Wait for all input to go

       ## Now do some more I/O.
       $in = "second input\n";
       pump $h until $out =~ /second input/;

       ## Clean up
       finish $h or die "cat returned $?";

    You can optionally compile the harness with harness() prior to
    start()ing or run()ing, and you may omit start() between harness() and
    pump(). You might want to do these things if you compile your harnesses
    ahead of time.

  Using regexps to match output
    As shown in most of the scripting examples, the read-to-scalar facility
    for gathering subcommand's output is often used with regular expressions
    to detect stopping points. This is because subcommand output often
    arrives in dribbles and drabs, often only a character or line at a time.
    This output is input for the main program and piles up in variables like
    the $out and $err in our examples.

    Regular expressions can be used to wait for appropriate output in
    several ways. The "cat" example in the previous section demonstrates how
    to pump() until some string appears in the output. Here's an example
    that uses "smb" to fetch files from a remote server:

       $h = harness \@smbclient, \$in, \$out;

       $in = "cd /src\n";
       $h->pump until $out =~ /^smb.*> \Z/m;
       die "error cding to /src:\n$out" if $out =~ "ERR";
       $out = '';

       $in = "mget *\n";
       $h->pump until $out =~ /^smb.*> \Z/m;
       die "error retrieving files:\n$out" if $out =~ "ERR";

       $in = "quit\n";
       $h->finish;

    Notice that we carefully clear $out after the first command/response
    cycle? That's because IPC::Run does not delete $out when we continue,
    and we don't want to trip over the old output in the second
    command/response cycle.

    Say you want to accumulate all the output in $out and analyze it
    afterwards. Perl offers incremental regular expression matching using
    the "m//gc" and pattern matching idiom and the "\G" assertion. IPC::Run
    is careful not to disturb the current "pos()" value for scalars it
    appends data to, so we could modify the above so as not to destroy $out
    by adding a couple of "/gc" modifiers. The "/g" keeps us from tripping
    over the previous prompt and the "/c" keeps us from resetting the prior
    match position if the expected prompt doesn't materialize immediately:

       $h = harness \@smbclient, \$in, \$out;

       $in = "cd /src\n";
       $h->pump until $out =~ /^smb.*> \Z/mgc;
       die "error cding to /src:\n$out" if $out =~ "ERR";

       $in = "mget *\n";
       $h->pump until $out =~ /^smb.*> \Z/mgc;
       die "error retrieving files:\n$out" if $out =~ "ERR";

       $in = "quit\n";
       $h->finish;

       analyze( $out );

    When using this technique, you may want to preallocate $out to have
    plenty of memory or you may find that the act of growing $out each time
    new input arrives causes an "O(length($out)^2)" slowdown as $out grows.
    Say we expect no more than 10,000 characters of input at the most. To
    preallocate memory to $out, do something like:

       my $out = "x" x 10_000;
       $out = "";

    "perl" will allocate at least 10,000 characters' worth of space, then
    mark the $out as having 0 length without freeing all that yummy RAM.

  Timeouts and Timers
    More than likely, you don't want your subprocesses to run forever, and
    sometimes it's nice to know that they're going a little slowly. Timeouts
    throw exceptions after a some time has elapsed, timers merely cause
    pump() to return after some time has elapsed. Neither is reset/restarted
    automatically.

    Timeout objects are created by calling timeout( $interval ) and passing
    the result to run(), start() or harness(). The timeout period starts
    ticking just after all the child processes have been fork()ed or
    spawn()ed, and are polled for expiration in run(), pump() and finish().
    If/when they expire, an exception is thrown. This is typically useful to
    keep a subprocess from taking too long.

    If a timeout occurs in run(), all child processes will be terminated and
    all file/pipe/ptty descriptors opened by run() will be closed. File
    descriptors opened by the parent process and passed in to run() are not
    closed in this event.

    If a timeout occurs in pump(), pump_nb(), or finish(), it's up to you to
    decide whether to kill_kill() all the children or to implement some more
    graceful fallback. No I/O will be closed in pump(), pump_nb() or
    finish() by such an exception (though I/O is often closed down in those
    routines during the natural course of events).

    Often an exception is too harsh. timer( $interval ) creates timer
    objects that merely prevent pump() from blocking forever. This can be
    useful for detecting stalled I/O or printing a soothing message or "."
    to pacify an anxious user.

    Timeouts and timers can both be restarted at any time using the timer's
    start() method (this is not the start() that launches subprocesses). To
    restart a timer, you need to keep a reference to the timer:

       ## Start with a nice long timeout to let smbclient connect.  If
       ## pump or finish take too long, an exception will be thrown.

     my $h;
     eval {
       $h = harness \@smbclient, \$in, \$out, \$err, ( my $t = timeout 30 );
       sleep 11;  # No effect: timer not running yet

       start $h;
       $in = "cd /src\n";
       pump $h until ! length $in;

       $in = "ls\n";
       ## Now use a short timeout, since this should be faster
       $t->start( 5 );
       pump $h until ! length $in;


Changes
Revision history for Perl extension IPC::Run

1.00 ???
 - #99 - Fix using fd in child process when it happens to be the same number in
   the child as it was in the parent.

0.99 Fri Mar 30 2018
 - Fixes for windows unit tests so they skip or pass.
  t/autoflush.t
  t/readonly.t
  t/run_stdin-callback-return-array.t

0.98 Thu Mar 29 2018
 - Switch to using $!{EINTR}
 - Honor umask for user/group permissions on a +rw file
 - Remove commented code from RT 50739
 - #63 - Patch for fixing _write on invalid objects
 - #54 - Make slave the controlling terminal so more programs can be called from IPC::Run reliably
 - #53 - Prevent Not a GLOB reference in close_terminal by untieing before close
 - #46 - '&' closing the stdin of the right process.
 - #41 - Do POSIX::_exit not ::exit
 - #40 - Assure child Win32 processes have same priority as parent
 - #38 - Add unit test for passing use of Readonly.pm
 - #39 - Fix GLOB test in IPC::Run::IO->new
 - #86 - Fix for memory leak [rt.cpan.org #57990]
 - MANIFEST was updated. New tests may start shipping now.

0.97 Mon Mar 26 2018
 - Update meta and POD to point to GitHub issues not RT.
 - Update README with pod changes.
 - Fix variable name in synopsis - [Github #47]
 - Fix t/run.t to work on "perl in space" - [Github #100]

0.96 Fri May 12 2017
 - Update bug tracker to https://github.com/toddr/IPC-Run/issues

0.95 Mon Apr 24 2017
 - Spelling fixes
 - doc fix: full_result returns $?, not $? >> 8
 - Fix check for reaped PID in reap_nb
 - Re-remove unnecessary use of UNIVERSAL

0.94_02 Thu Apr 14 2016
 - Skip infinity test on perl below 5.8.9

0.94_01 Tue Apr 12 2016
 - Enable Travis-CI integration for IPC::Run
 - Use q{} and qq{} to fix problems with quotation in cmd.exe
 - sync $VERSION across all files to address kwalitee
 - RT 109243 - Stop side affect of altering flushing of STDOUT and STDERR
 - CHILD_HANDLE should be closed with 'Close' to prevent a warning.
 - require Win32API::File on Windows
 - Better timer error message + an additional unit test
 - Catching previously non-detected malformed time strings
 - Let Timer accept all allowable perl numbers
 - allow the OS to choose the ephemeral port to use
 - Don't use version.pm to parse the perl version in Makefile.PL
 - perltidy
 - Do not import POSIX into local namespace (it's a memory hog)

0.94 Sun Dec 14 2014
 - RT 100858 - Update License discrepancies.

0.93_01 Wed Dec 10 2014
 - Many typo corrections from David Steinbrunner via github
 - RT 20105 - t/pty.t fails on Solaris 10

0.93 Wed Dec 10 2014
 - RT 100711 - Add aggressive exit to t/signal.t after skip_all to prevent failures.
 - RT 97926 - Fix for run argument with spaces in path
 - skip t/win32_compile.t on Android
 - RT 86133 - TODO failing POD tests for IPC::Run

0.92 Thu Aug 30 2012
 - Tests are good in dev version. Releasing to stable.

0.92_01 Wed 22 Aug 2012
 - RT 59775 - Catching a signal during select() can lock up IPC::Run

0.91 Tue 14 Feb 2012
 - Tests are good in dev version. Releasing to stable.

0.91_01 Mon 16 Jan 2012
 - RT 57277 - Apply suggested 1 line fix for dev release to public.

0.90 Wed 29 Jun 2011
 - RT 57277 - Add breaking test. No solution at present.
 - RT 68623 - disable parallel tests on windows
 - RT 43072 - Skip dragonfly platform on t/pty.t (hangs)
 - RT 14078 - run() miscalculates length of UTF-8 strings
 - RT 49693 - security patch for lib/IPC/Run/Win32IO.pm s/INADDR_ANY//
 - RT 38193 - Fix error detection on non-English operating systems
 - Add a blank doc entry for IPC::Run::Win32IO::poll to quiet windows pod parser errors
 - RT 57190 - handle PATH environment variable not being defined. WILL NOT default to '.' since UNIX would not do this and the program has been behaving the UNIX way to date.

0.89 Wed 31 Mar 2010
  - Revert extended tests to require "oslo consensus" variables - http://use.perl.org/~Alias/journal/36128
  - Add IO::Pty to META.yml requirement This has been inconsistently depended on over the years in META.yml

0.88 Tue 30 Mar 2010
  - Missed dangling isa in IPC::Run::Debug
  - Fix retry for read of fh when I get "Resource temporarily unavailable."
    It was previously pounding on the file handle with no rest 200 times then giving up.
    On fast systems this wasn't long enough. I'm implementing select(undef, undef, 0.01)
    to provide a more consistent retry among hardware and os.
  - Fix POD indentation in IO.pm

0.87 Mon 29 Mar 2010
  - Add doc type fixes - RT 56030 from carnil (Debian)
  - Remove recommends for AUTHOR only modules
  - add pod_coverage and simplify pod test.
  - document undocumented public subs with TODO and best guess as to what they do.
  - Explicitly recommend IO::Pty version number 1.08

0.86 Wed 24 Mar 2010
  - Add all files to subversion for easier releases
  - bump version on ALL files to 86

0.85 Mon 22 Mar 2010
  - Remove use UNIVERSAL. - RT 51509
  - fix pod for Win32Helper - RT 51695
  - Better PATHEXT logic bug if cmd_name contains periods - RT 50739
  - Enhance tests for run.t for better diagnostics - RT 47630

0.84 Mon 13 Jul 2009
	- Add darwin to the list of known-bad platforms

0.83 Fri 10 Jul 2009
	- Switch from command.com to cmd.exe so we work on 2008 or newer.

0.82 Thu 18 Dec 2008
	- Moving changes in 0.81_01 to a production release

0.81_01 Wed 15 Oct 2008
	- This is the first in a series of refactoring test releases.
	- Removed Makefile.PL message noisily asking users to do CPAN Testers
	  dirty work.
	- Simplified the Makefile.PL code a little.
	- Upgraded all tests to Test::More
	- Added a $VERSION for all modules
	- Adding some missing POD sections
	- Various other clean ups

0.80 (missing)
	- IPC::Run::IO now retries on certain "temporarily unavailable" errors.
	  This should fix several reported issues with t/run.t, test 69.
	  Many thanks to < Eric (at) Scratch Computing (.com) > for the patch!
	- Applied documentation patch from RT.
	- Fixed documentation to work with '<' redirect

0.79 Wed Jan 19 15:39:00 PST 2005

lib/IPC/Run/Win32Helper.pm
package IPC::Run::Win32Helper;

=pod

=head1 NAME

IPC::Run::Win32Helper - helper routines for IPC::Run on Win32 platforms.

=head1 SYNOPSIS

    use IPC::Run::Win32Helper;   # Exports all by default

=head1 DESCRIPTION

IPC::Run needs to use sockets to redirect subprocess I/O so that the select() loop
will work on Win32. This seems to only work on WinNT and Win2K at this time, not
sure if it will ever work on Win95 or Win98. If you have experience in this area, please
contact me at barries@slaysys.com, thanks!.

=cut

use strict;
use Carp;
use IO::Handle;
use vars qw{ $VERSION @ISA @EXPORT };

BEGIN {
    $VERSION = '0.98';
    @ISA     = qw( Exporter );
    @EXPORT  = qw(
      win32_spawn
      win32_parse_cmd_line
      _dont_inherit
      _inherit
    );
}

require POSIX;

use Text::ParseWords;
use Win32::Process;
use IPC::Run::Debug;
use Win32API::File qw(
  FdGetOsFHandle
  SetHandleInformation
  HANDLE_FLAG_INHERIT
  INVALID_HANDLE_VALUE
);

## Takes an fd or a GLOB ref, never never never a Win32 handle.
sub _dont_inherit {
    for (@_) {
        next unless defined $_;
        my $fd = $_;
        $fd = fileno $fd if ref $fd;
        _debug "disabling inheritance of ", $fd if _debugging_details;
        my $osfh = FdGetOsFHandle $fd;
        croak $^E if !defined $osfh || $osfh == INVALID_HANDLE_VALUE;

        SetHandleInformation( $osfh, HANDLE_FLAG_INHERIT, 0 );
    }
}

sub _inherit {    #### REMOVE
    for (@_) {    #### REMOVE
        next unless defined $_;    #### REMOVE
        my $fd = $_;               #### REMOVE
        $fd = fileno $fd if ref $fd;    #### REMOVE
        _debug "enabling inheritance of ", $fd if _debugging_details;    #### REMOVE
        my $osfh = FdGetOsFHandle $fd;                                   #### REMOVE
        croak $^E if !defined $osfh || $osfh == INVALID_HANDLE_VALUE;    #### REMOVE
        #### REMOVE
        SetHandleInformation( $osfh, HANDLE_FLAG_INHERIT, 1 );           #### REMOVE
    }    #### REMOVE
}    #### REMOVE
#### REMOVE
#sub _inherit {
#   for ( @_ ) {
#      next unless defined $_;
#      my $osfh = GetOsFHandle $_;
#      croak $^E if ! defined $osfh || $osfh == INVALID_HANDLE_VALUE;
#      SetHandleInformation( $osfh, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT );
#   }
#}

=pod

=head1 FUNCTIONS

=over

=item optimize()

Most common incantations of C<run()> (I<not> C<harness()>, C<start()>,
or C<finish()>) now use temporary files to redirect input and output
instead of pumper processes.

Temporary files are used when sending to child processes if input is
taken from a scalar with no filter subroutines.  This is the only time
we can assume that the parent is not interacting with the child's
redirected input as it runs.

Temporary files are used when receiving from children when output is
to a scalar or subroutine with or without filters, but only if
the child in question closes its inputs or takes input from 
unfiltered SCALARs or named files.  Normally, a child inherits its STDIN
from its parent; to close it, use "0<&-" or the C<< noinherit => 1 >> option.
If data is sent to the child from CODE refs, filehandles or from
scalars through filters than the child's outputs will not be optimized
because C<optimize()> assumes the parent is interacting with the child.
It is ok if the output is filtered or handled by a subroutine, however.

This assumes that all named files are real files (as opposed to named
pipes) and won't change; and that a process is not communicating with
the child indirectly (through means not visible to IPC::Run).
These can be an invalid assumptions, but are the 99% case.
Write me if you need an option to enable or disable optimizations; I
suspect it will work like the C<binary()> modifier.

To detect cases that you might want to optimize by closing inputs, try
setting the C<IPCRUNDEBUG> environment variable to the special C<notopt>
value:

   C:> set IPCRUNDEBUG=notopt
   C:> my_app_that_uses_IPC_Run.pl

=item optimizer() rationalizations

Only for that limited case can we be sure that it's ok to batch all the
input in to a temporary file.  If STDIN is from a SCALAR or from a named
file or filehandle (again, only in C<run()>), then outputs to CODE refs
are also assumed to be safe enough to batch through a temp file,
otherwise only outputs to SCALAR refs are batched.  This can cause a bit
of grief if the parent process benefits from or relies on a bit of
"early returns" coming in before the child program exits.  As long as
the output is redirected to a SCALAR ref, this will not be visible.
When output is redirected to a subroutine or (deprecated) filters, the
subroutine will not get any data until after the child process exits,
and it is likely to get bigger chunks of data at once.

The reason for the optimization is that, without it, "pumper" processes
are used to overcome the inconsistencies of the Win32 API.  We need to
use anonymous pipes to connect to the child processes' stdin, stdout,
and stderr, yet select() does not work on these.  select() only works on
sockets on Win32.  So for each redirected child handle, there is
normally a "pumper" process that connects to the parent using a
socket--so the parent can select() on that fd--and to the child on an
anonymous pipe--so the child can read/write a pipe.

Using a socket to connect directly to the child (as at least one MSDN
article suggests) seems to cause the trailing output from most children
to be lost.  I think this is because child processes rarely close their
stdout and stderr explicitly, and the winsock dll does not seem to flush
output when a process that uses it exits without explicitly closing
them.

Because of these pumpers and the inherent slowness of Win32
CreateProcess(), child processes with redirects are quite slow to
launch; so this routine looks for the very common case of
reading/writing to/from scalar references in a run() routine and
converts such reads and writes in to temporary file reads and writes.

Such files are marked as FILE_ATTRIBUTE_TEMPORARY to increase speed and
as FILE_FLAG_DELETE_ON_CLOSE so it will be cleaned up when the child
process exits (for input files).  The user's default permissions are
used for both the temporary files and the directory that contains them,
hope your Win32 permissions are secure enough for you.  Files are
created with the Win32API::File defaults of
FILE_SHARE_READ|FILE_SHARE_WRITE.

Setting the debug level to "details" or "gory" will give detailed
information about the optimization process; setting it to "basic" or
higher will tell whether or not a given call is optimized.  Setting
it to "notopt" will highlight those calls that aren't optimized.

=cut

sub optimize {
    my ($h) = @_;

    my @kids = @{ $h->{KIDS} };

    my $saw_pipe;

    my ( $ok_to_optimize_outputs, $veto_output_optimization );

    for my $kid (@kids) {
        ( $ok_to_optimize_outputs, $veto_output_optimization ) = ()
          unless $saw_pipe;

        _debug "Win32 optimizer: (kid $kid->{NUM}) STDIN piped, carrying over ok of non-SCALAR output optimization"
          if _debugging_details && $ok_to_optimize_outputs;
        _debug "Win32 optimizer: (kid $kid->{NUM}) STDIN piped, carrying over veto of non-SCALAR output optimization"
          if _debugging_details && $veto_output_optimization;

        if ( $h->{noinherit} && !$ok_to_optimize_outputs ) {
            _debug "Win32 optimizer: (kid $kid->{NUM}) STDIN not inherited from parent oking non-SCALAR output optimization"
              if _debugging_details && $ok_to_optimize_outputs;
            $ok_to_optimize_outputs = 1;
        }

        for ( @{ $kid->{OPS} } ) {
            if ( substr( $_->{TYPE}, 0, 1 ) eq "<" ) {
                if ( $_->{TYPE} eq "<" ) {
                    if ( @{ $_->{FILTERS} } > 1 ) {
                        ## Can't assume that the filters are idempotent.
                    }
                    elsif (ref $_->{SOURCE} eq "SCALAR"
                        || ref $_->{SOURCE} eq "GLOB"
                        || UNIVERSAL::isa( $_, "IO::Handle" ) ) {
                        if ( $_->{KFD} == 0 ) {
                            _debug
                              "Win32 optimizer: (kid $kid->{NUM}) 0$_->{TYPE}",
                              ref $_->{SOURCE},
                              ", ok to optimize outputs"
                              if _debugging_details;
                            $ok_to_optimize_outputs = 1;
                        }
                        $_->{SEND_THROUGH_TEMP_FILE} = 1;
                        next;
                    }
                    elsif ( !ref $_->{SOURCE} && defined $_->{SOURCE} ) {
                        if ( $_->{KFD} == 0 ) {
                            _debug
                              "Win32 optimizer: (kid $kid->{NUM}) 0<$_->{SOURCE}, ok to optimize outputs",
                              if _debugging_details;
                            $ok_to_optimize_outputs = 1;
                        }
                        next;
                    }
                }
                _debug
                  "Win32 optimizer: (kid $kid->{NUM}) ",
                  $_->{KFD},
                  $_->{TYPE},
                  defined $_->{SOURCE}
                  ? ref $_->{SOURCE}
                      ? ref $_->{SOURCE}
                      : $_->{SOURCE}
                  : defined $_->{FILENAME} ? $_->{FILENAME}
                  : "",
                  @{ $_->{FILTERS} } > 1 ? " with filters" : (),
                  ", VETOING output opt."
                  if _debugging_details || _debugging_not_optimized;
                $veto_output_optimization = 1;
            }
            elsif ( $_->{TYPE} eq "close" && $_->{KFD} == 0 ) {
                $ok_to_optimize_outputs = 1;
                _debug "Win32 optimizer: (kid $kid->{NUM}) saw 0<&-, ok to optimize outputs"
                  if _debugging_details;
            }
            elsif ( $_->{TYPE} eq "dup" && $_->{KFD2} == 0 ) {
                $veto_output_optimization = 1;
                _debug "Win32 optimizer: (kid $kid->{NUM}) saw 0<&$_->{KFD2}, VETOING output opt."
                  if _debugging_details || _debugging_not_optimized;
            }
            elsif ( $_->{TYPE} eq "|" ) {
                $saw_pipe = 1;
            }
        }

        if ( !$ok_to_optimize_outputs && !$veto_output_optimization ) {
            _debug "Win32 optimizer: (kid $kid->{NUM}) child STDIN not redirected, VETOING non-SCALAR output opt."
              if _debugging_details || _debugging_not_optimized;
            $veto_output_optimization = 1;
        }

        if ( $ok_to_optimize_outputs && $veto_output_optimization ) {
            $ok_to_optimize_outputs = 0;
            _debug "Win32 optimizer: (kid $kid->{NUM}) non-SCALAR output optimizations VETOed"
              if _debugging_details || _debugging_not_optimized;
        }

        ## SOURCE/DEST ARRAY means it's a filter.
        ## TODO: think about checking to see if the final input/output of
        ## a filter chain (an ARRAY SOURCE or DEST) is a scalar...but
        ## we may be deprecating filters.

        for ( @{ $kid->{OPS} } ) {
            if ( $_->{TYPE} eq ">" ) {
                if (
                    ref $_->{DEST} eq "SCALAR"
                    || (
                        (
                               @{ $_->{FILTERS} } > 1
                            || ref $_->{DEST} eq "CODE"
                            || ref $_->{DEST} eq "ARRAY"    ## Filters?
                        )
                        && ( $ok_to_optimize_outputs && !$veto_output_optimization )
                    )
                  ) {
                    $_->{RECV_THROUGH_TEMP_FILE} = 1;
                    next;
                }
                _debug
                  "Win32 optimizer: NOT optimizing (kid $kid->{NUM}) ",
                  $_->{KFD},
                  $_->{TYPE},
                  defined $_->{DEST}
                  ? ref $_->{DEST}
                      ? ref $_->{DEST}
                      : $_->{SOURCE}
                  : defined $_->{FILENAME} ? $_->{FILENAME}
                  : "",
                  @{ $_->{FILTERS} } ? " with filters" : (),
                  if _debugging_details;
            }
        }
    }

}

=pod

=item win32_parse_cmd_line

   @words = win32_parse_cmd_line( q{foo bar 'baz baz' "bat bat"} );

returns 4 words. This parses like the bourne shell (see
the bit about shellwords() in L<Text::ParseWords>), assuming we're
trying to be a little cross-platform here.  The only difference is
that "\" is *not* treated as an escape except when it precedes 
punctuation, since it's used all over the place in DOS path specs.

TODO: globbing? probably not (it's unDOSish).

TODO: shebang emulation? Probably, but perhaps that should be part
of Run.pm so all spawned processes get the benefit.

LIMITATIONS: shellwords dies silently on malformed input like 

   a\"

=cut

sub win32_parse_cmd_line {
    my $line = shift;
    $line =~ s{(\\[\w\s])}{\\$1}g;
    return shellwords $line;
}

=pod

=item win32_spawn

Spawns a child process, possibly with STDIN, STDOUT, and STDERR (file descriptors 0, 1, and 2, respectively) redirected.

B<LIMITATIONS>.

Cannot redirect higher file descriptors due to lack of support for this in the
Win32 environment.

This can be worked around by marking a handle as inheritable in the
parent (or leaving it marked; this is the default in perl), obtaining it's
Win32 handle with C<Win32API::GetOSFHandle(FH)> or
C<Win32API::FdGetOsFHandle($fd)> and passing it to the child using the command
line, the environment, or any other IPC mechanism (it's a plain old integer).
The child can then use C<OsFHandleOpen()> or C<OsFHandleOpenFd()> and possibly
C<<open FOO ">&BAR">> or C<<open FOO ">&$fd>> as need be.  Ach, the pain!

Remember to check the Win32 handle against INVALID_HANDLE_VALUE.

=cut

sub _save {
    my ( $saved, $saved_as, $fd ) = @_;

    ## We can only save aside the original fds once.
    return if exists $saved->{$fd};

    my $saved_fd = IPC::Run::_dup($fd);
    _dont_inherit $saved_fd;

    $saved->{$fd}          = $saved_fd;
    $saved_as->{$saved_fd} = $fd;

    _dont_inherit $saved->{$fd};
}

sub _dup2_gently {
    my ( $saved, $saved_as, $fd1, $fd2 ) = @_;
    _save $saved, $saved_as, $fd2;

    if ( exists $saved_as->{$fd2} ) {
        ## The target fd is colliding with a saved-as fd, gotta bump
        ## the saved-as fd to another fd.
        my $orig_fd  = delete $saved_as->{$fd2};
        my $saved_fd = IPC::Run::_dup($fd2);
        _dont_inherit $saved_fd;

        $saved->{$orig_fd}     = $saved_fd;
        $saved_as->{$saved_fd} = $orig_fd;
    }
    _debug "moving $fd1 to kid's $fd2" if _debugging_details;
    IPC::Run::_dup2_rudely( $fd1, $fd2 );
}

sub win32_spawn {
    my ( $cmd, $ops ) = @_;

    ## NOTE: The debug pipe write handle is passed to pump processes as STDOUT.
    ## and is not to the "real" child process, since they would not know
    ## what to do with it...unlike Unix, we have no code executing in the
    ## child before the "real" child is exec()ed.

    my %saved;       ## Map of parent's orig fd -> saved fd
    my %saved_as;    ## Map of parent's saved fd -> orig fd, used to
    ## detect collisions between a KFD and the fd a
    ## parent's fd happened to be saved to.

    for my $op (@$ops) {
        _dont_inherit $op->{FD} if defined $op->{FD};

        if ( defined $op->{KFD} && $op->{KFD} > 2 ) {
            ## TODO: Detect this in harness()
            ## TODO: enable temporary redirections if ever necessary, not
            ## sure why they would be...
            ## 4>&1 1>/dev/null 1>&4 4>&-
            croak "Can't redirect fd #", $op->{KFD}, " on Win32";
        }

        ## This is very similar logic to IPC::Run::_do_kid_and_exit().
        if ( defined $op->{TFD} ) {
            unless ( $op->{TFD} == $op->{KFD} ) {
                _dup2_gently \%saved, \%saved_as, $op->{TFD}, $op->{KFD};
                _dont_inherit $op->{TFD};
            }
        }
        elsif ( $op->{TYPE} eq "dup" ) {
            _dup2_gently \%saved, \%saved_as, $op->{KFD1}, $op->{KFD2}
              unless $op->{KFD1} == $op->{KFD2};
        }
        elsif ( $op->{TYPE} eq "close" ) {
            _save \%saved, \%saved_as, $op->{KFD};
            IPC::Run::_close( $op->{KFD} );
        }
        elsif ( $op->{TYPE} eq "init" ) {
            ## TODO: detect this in harness()
            croak "init subs not allowed on Win32";
        }
    }

    my $process;
    my $cmd_line = join " ", map {
        ( my $s = $_ ) =~ s/"/"""/g;
        $s = qq{"$s"} if /[\"\s]|^$/;
        $s;
    } @$cmd;

    _debug "cmd line: ", $cmd_line
      if _debugging;

    Win32::Process::Create(
        $process,
        $cmd->[0],
        $cmd_line,
        1,    ## Inherit handles
        0,    ## Inherit parent priortiy class. Was NORMAL_PRIORITY_CLASS
        ".",
    ) or croak "$!: Win32::Process::Create()";

    for my $orig_fd ( keys %saved ) {
        IPC::Run::_dup2_rudely( $saved{$orig_fd}, $orig_fd );
        IPC::Run::_close( $saved{$orig_fd} );
    }

    return ( $process->GetProcessID(), $process );
}

1;

=pod

=back

=head1 AUTHOR

Barries Slaymaker <barries@slaysys.com>.  Funded by Perforce Software, Inc.

=head1 COPYRIGHT

Copyright 2001, Barrie Slaymaker, All Rights Reserved.

You may use this under the terms of either the GPL 2.0 or the Artistic License.

=cut
